jane/doc/extensions/uniqueness/intro.md

```diff
@@ -0,0 +1,123 @@
+# Introduction to Uniqueness
+
+See also the full feature [reference](reference.md) and [common pitfalls](pitfalls.md).
+In this document, we use the new [syntax for modes](../modes/syntax.md),
+which we expect to be stable soonish.
+
+The `unique` mode designates values that have only a single reference pointing
+to them. For example, an externally allocated data structure might support a
+`free` operation:
```

Comment:
Hmm how about something like:
> The `unique` mode designates a reference to a value that is the only reference to that value throughout the program.

Like, the `unique` is on the reference itself, not the value it points to.

Reply:
I do think that uniqueness is on the value and find the current phrasing clearer than the proposed one. The subtle special case where you can use a value uniquely in one branch and share it in the other still makes sense thinking about references instead of usages and is only explained in the `reference.md` document.

Reply:
In particular, I would want readers to form a denotational semantics in their head here. "This value has only one reference" is such a semantics, while "This values usage in the program is the only usage" is more operational.

Reply:
Consider the example:
```
let x = ... in
if b then use_as_aliased x
else use_as_unique x
```
we have some `x` in the environment, and some `Pexp_ident x` in the expression. The former is a "value" while the later is a "usage/reference" of that "value. My point is that, it's often more useful to talk about uniqueness for the latter, not the former. In this example, there is only one value, but it's `unique` sometimes, and `shared` sometimes?

Reply:
I am not saying that `Pexp_ident x` is itself a value and I realized that you were thinking about the `unique_use` trick. But I still think it is helpful to think of modes using the denotational semantics of what it means when a value has that mode. 

This is also how we teach locality in the `local/intro.md` file: `These [local] values live on a separate stack`. Following your argument, we should expect users to be confused that a `global` value may be used as `global` sometimes and as `local` another time. Would that document be clearer if we wrote: "The `local` mode designates a usage of a value that the value may not escape from"? No, because the escape-restriction is just an incidental restriction that is necessary to ensure the soundness of the denotational meaning. Similarly here, the restriction that the value may only have one unique use in the control flow establishes the denotational meaning that the value only has one reference pointing to it.

Reply:
I think a constrast that was interesting to me was that, `unique_use` contains `Uniqueness.r` for uniqueness, but `Linearity.l` for linearity. This is probably hinting at the duality between monadic and comonadic axes. Locality being comonadic probably doesn't fall within my previous argument.

> Following your argument, we should expect users to be confused that a global value may be used as global sometimes and as local another time.

If we have a `global` variable in the environment, and we use it as `local` somewhere, the program will continue to type check if that usage is strenghthened to `global`. Therefore, the "global-ness` seems to be a property of the variable/value that we can use at will.

The same can't be said for monadic axes. If we have a `uncontended` variable in the environment, and we use it as `contended` somewhere, the program might fail to type check if the usage is strengthenened to `uncontended`. Here it's probably more useful to talk about the contention of the two usages, rather than that of the bound variable.

Reply:
Right, the uniqueness and contention of a value changes as time goes on. That means that you often reason backwards: this operation expects a unique value so you look at the code before it to see if the value is still unique. You ensure that the value has only one reference at this point by making sure that there is no use coming before it.

FWIW, the opening paragraph now reads:

> The `unique` mode designates values that have only a single reference pointing
to them. If an operation takes a `unique` argument, it will consume the only
reference to the value. For example, an externally allocated data structure
might support a `free` operation:

Are you okay with that phrasing? What would you write instead?

Reply:
> That means that you often reason backwards

Yes I guess the backwardness is what makes me prefer "usage" over "value", as a value goes forward only: a global value can always be used as global, no matter what.

I still prefer to say "unique designates usages". But in the end it's just a matter of English which is subjective. So I'm happy to resolve.